### ISSUE TITLE
Issues:  self.action_grads = tf.gradients(self.out, self.action)

### ISSUE THREAD

**Issue by cxfneo** [pemami4911/deep-rl#3]
HI 
    see the link first :https://github.com/tensorflow/tensorflow/issues/675
    if self.out is a vector and self.action is a vector ,  the result should be a Jacobian matrix , but "Currently if you call gradients(ys, xs), it will return the sum of dy/dx over all ys for each x in xs."
So I think the expression used in your code is not the one as written in DDPG paper.

    Thank u very much

**Comment by pemami4911** [comment]
Hi,

`self.out` is actually a Tensor of shape [MINIBATCH_SIZE, 1] and `self.action` is a Tensor of shape `[MINIBATCH_SIZE, self.action_dim]`. Calling `tf.gradients(self.out, self.action)` is correct; for each action in the minibatch (i.e., for each x in xs), this will sum up the gradients of each critic output in the minibatch w.r.t. that action (i.e., sum of dy/dx over all ys). **There is an error here though**, in that I should be dividing the result by the size of the minibatch. I'm not sure why Tensorflow doesn't do this automatically. I'll add that to my to-do list, but feel free to submit a pull-request with a fix - it should be quick.

Thanks!

**Comment by pemami4911** [comment]
Resolved in commit bf21e10d6a977a15fd96fe38d80e2294cf5f9bdc

**Comment by abred** [comment]
isn't there still an error?
```
self.out.shape = [MINIBATCH_SIZE, 1]
self.action.shape = [MINIBATCH_SIZE, self.action_dim]
```
let's assume MINIBATCH_SIZE = 4
```
self.out =
  [[o11],
   [o21],
   [o31],
   [o41]]

self.action =
  [[a11, a12, a13],
   [a21, a22, a23],
   [a31, a32, a33],
   [a41, a42, a43]]

ys = self.out
xs = self.action
tf.gradients(ys, xs) -> sum(dy/dx) for y in ys

g11 = o11/da11 + o21/da11 + o31/da11 + o41/da11
g12 = o11/da12 + o21/da12 + o31/da12 + o41/da12
...
g43 = o11/da43 + o21/da43 + o31/da43 + o41/da43

grads = 
  [[g11, g12, g13],
   [g21, g22, g23],
   [g31, g32, g33],
   [g41, g42, g41]]
grads = tf.div(grads, MINIBATCH_SIZE)

actor.train(s_batch, grads[0])  -> only [g11, g12, g13]
```
so shouldn't it be 
```
grads = tf.reduce_mean(grads, 0)
```

**Comment by abred** [comment]
(changed a bit)
isn't there still an error?
```
self.out.shape = [MINIBATCH_SIZE, 1]
self.action.shape = [MINIBATCH_SIZE, self.action_dim]
```
let's assume MINIBATCH_SIZE = 4
so we have 4 state-action pairs (si,ai)
```
self.out = critic([s1, s2, s3, s4])
self.out  ->
  [[o1],
   [o2],
   [o3],
   [o4]]

self.action =
  [[a11, a12, a13],
   [a21, a22, a23],
   [a31, a32, a33],
   [a41, a42, a43]]

ys = self.out
xs = self.action
tf.gradients(ys, xs) -> sum(dy/dx) for y in ys

g11 = o1/da11 + o2/da11 + o3/da11 + o4/da11
g12 = o1/da12 + o2/da12 + o3/da12 + o4/da12
...
g43 = o1/da43 + o2/da43 + o3/da43 + o4/da43

grads = 
[[g11, g12, g13],
 [g21, g22, g23],
 [g31, g32, g33],
 [g41, g42, g41]]
grads = tf.div(grads, MINIBATCH_SIZE)
```
but don't we only want the derivative of some output wrt. the matching action (instead of the sum)?
```
g11 = o1/da11 (->critic(s1)/da11)
g12 = o1/da12 (->critic(s1)/da12)
g21 = o2/da21 (->critic(s2)/da12)
...
g43 = o4/da43 (->critic(s4)/da43)
```

**Comment by pemami4911** [comment]
Note that the action dimension is just 1.  I see the issue you're talking about though; I'm trying to figure out a way to do 

```
g1 = o1/da1 (->critic(s1)/da1)
g2 = o2/da2 (->critic(s2)/da2)
g3 = o3/da3 (->critic(s3)/da3)
...
g64 = o64/da64 (->critic(s64)/da64)
```
without calling `tf.gradients` MINIBATCH_SIZE (e.g., 64) times each update. I'm looking at https://arxiv.org/pdf/1510.01799.pdf and https://github.com/tensorflow/tensorflow/issues/4897 but haven't figured out a solution for this specific situation yet. Any thoughts?

**Comment by abred** [comment]
haven't tried it yet, but what happens if we change out to:
```
self.out  ->
  [[o1, 0, 0, 0],
   [0, o2, 0, 0],
   [0, 0, o3, 0],
   [0, 0, 0, o4]]
```

**Comment by mlogic** [comment]
It would be great if the code can be general and support multiple actions. I've done a little test based on @abred 's notion above:
```python
import tensorflow as tf

action = tf.placeholder(tf.float32, [None, 3])
out_op = tf.div(action, tf.constant(2.0))
grad_op = tf.gradients(out_op, action)

sess = tf.Session()
out, grad = sess.run([out_op, grad_op], {action: [[1,2,3], [4,5,6], [7,8,9], [10,11,12]]})

print(out)
print(grad)
```
and here's output
```
[[ 0.5  1.   1.5]
 [ 2.   2.5  3. ]
 [ 3.5  4.   4.5]
 [ 5.   5.5  6. ]]
[array([[ 0.5,  0.5,  0.5],
       [ 0.5,  0.5,  0.5],
       [ 0.5,  0.5,  0.5],
       [ 0.5,  0.5,  0.5]], dtype=float32)]
```
So it seems to me that `gi,j = oi/dai,j` is just what we want. And this is contradicting TF documentation: they are not summed over samples in the minibatch. Or did I do anything wrong?

**Comment by pemami4911** [comment]
Check this out: 

```Python
import tensorflow as tf

MINIBATCH_SIZE = 4

action = tf.placeholder(tf.float32, [None, 3])

out_op1 = [tf.square(action[0]) for _ in range(MINIBATCH_SIZE)]
out_op2 = tf.square(action[0])
grad_op1 = tf.gradients(out_op1, action)
grad_op2 = tf.gradients(out_op2, action)

sess = tf.Session()

out1, out2, grad1, grad2 = sess.run([out_op1, out_op2, grad_op1, grad_op2], {action: [[1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12]]})

print("Out1: {}".format(out1))
print("Grad1: {}".format(grad1))
print("Out2: {}".format(out2))
print("Grad2: {}".format(grad2))

```
and the output is 

```
Out1: [array([ 1.,  4.,  9.], dtype=float32), array([ 1.,  4.,  9.], dtype=float32), array([ 1.,  4.,  9.], dtype=float32), array([ 1.,  4.,  9.], dtype=float32)]
Grad1: [array([[  8.,  16.,  24.],
       [  0.,   0.,   0.],
       [  0.,   0.,   0.],
       [  0.,   0.,   0.]], dtype=float32)]
Out2: [ 1.  4.  9.]
Grad2: [array([[ 2.,  4.,  6.],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.]], dtype=float32)]
```

So when you have a list of Tensors as your `ys`, `tf.gradients` computes the sum over each Tensor in the list. Hence, the difference between `Grad1` and `Grad2`. So a single Tensor of shape `[None, 3]` differentiated w.r.t. a single Tensor of shape `[None, 3]` won't do any averaging and just produces the Jacobian of interest. For our case, just calling `tf.gradients(self.out, self.action)` should be fine then! It actually doesn't even need to be divided by the batch size either. Sounds correct?

**Comment by mlogic** [comment]
Yes, that's what I'm thinking. We don't need that tf.div().

**Comment by abred** [comment]
It also works if out1 is tensor and not a list of tensors
```
import numpy as np
import tensorflow as tf

MINIBATCH_SIZE = 4

action = tf.placeholder(tf.float32, [4, 3])

out_op1 = tf.stack([tf.square(action[0]) for i in range(MINIBATCH_SIZE)])
out_op2 = tf.square(action)
grad_op1 = tf.gradients(out_op1, action)
grad_op2 = tf.gradients(out_op2, action)

sess = tf.Session()

ac = np.array([[1, 2, 3],
               [4, 5, 6],
               [7, 8, 9],
               [10, 11, 12.0]])
# ac = np.array([[1], [2], [3], [4]])
out1, out2, grad1, grad2 = sess.run(
    [out_op1,
     out_op2,
     grad_op1,
     grad_op2],
    {action: ac})

print("act: {}".format(ac))

print("Out1: {}".format(out1))
print("Grad1: {}".format(grad1))
print("Out2: {}".format(out2))
print("Grad2: {}".format(grad2))
```

I think we are lucky because we are computing the gradient with regard to an input.
tf.gradients computes `sum(dy/dx)`, `dx` is the input in our case. Each sample has a different input so `dyi/dxj = 0` if `i != j`
In general it is used to compute the gradient of a weight and  the weight is the same for each sample so the gradient is not necessarily zero

**Comment by pemami4911** [comment]
Okay, I see. So regardless of whether it is a list or not, it will accumulate gradients over a batch if the same inputs are used to produce the outputs. Interesting.

@abred what you are saying is correct; however, for `tf.gradients(self.out, self.action)`, `self.action` are the inputs. So we should be fine. See below:

For `Grad2`, the first row is `[6, 12, 18]`, because the partial derivative of `Out2_11`, `Out2_12`, and `Out2_13` w.r.t. actions other than those in the first "batch" (i.e., action[0]), is 0! So `g11 = dOut2_11/daction_11 + dOut2_12/daction_11 + dOut2_13/daction_11 =   (2 * action_11 * 1) + (2 * action_11 * 1) + (2 * action_11 * 1) = 6`. You can likewise derive the rest of `Grad2` this way. 

```Python
import numpy as np
import tensorflow as tf

MINIBATCH_SIZE = 4

action = tf.placeholder(tf.float32, [4, 3])

# weight matrix
params = tf.placeholder(tf.float32, [3, 3])
out_op1 = tf.stack([tf.matmul(tf.square(tf.reshape(action[0], [1, 3])), params) for i in range(MINIBATCH_SIZE)])
out_op2 = tf.matmul(tf.square(action), params)
grad_op1 = tf.gradients(out_op1, action)
grad_op2 = tf.gradients(out_op2, action)

sess = tf.Session()

ac = np.array([[1, 2, 3],
               [4, 5, 6],
               [7, 8, 9],
               [10, 11, 12.0]])

p = np.ones((3, 3))

# ac = np.array([[1], [2], [3], [4]])
out1, out2, grad1, grad2 = sess.run(
    [out_op1,
     out_op2,
     grad_op1,
     grad_op2],
    {action: ac, params: p})

print("act: {}".format(ac))
print("Out1: {}".format(out1))
print("Grad1: {}".format(grad1))
print("Out2: {}".format(out2))
print("Grad2: {}".format(grad2))
```

and output: 

```
act: [[  1.   2.   3.]
 [  4.   5.   6.]
 [  7.   8.   9.]
 [ 10.  11.  12.]]
Out1: [[[ 14.  14.  14.]]

 [[ 14.  14.  14.]]

 [[ 14.  14.  14.]]

 [[ 14.  14.  14.]]]
Grad1: [array([[ 24.,  48.,  72.],
       [  0.,   0.,   0.],
       [  0.,   0.,   0.],
       [  0.,   0.,   0.]], dtype=float32)]
Out2: [[  14.   14.   14.]
 [  77.   77.   77.]
 [ 194.  194.  194.]
 [ 365.  365.  365.]]
Grad2: [array([[  6.,  12.,  18.],
       [ 24.,  30.,  36.],
       [ 42.,  48.,  54.],
       [ 60.,  66.,  72.]], dtype=float32)]
```

Even though the same weights were used to compute the outputs for each minibatch, each batch of inputs only corresponds to one batch of outputs and so they won't get averaged over them. Does this make sense?

**Comment by pemami4911** [comment]
@abred if this is what you were already saying in your previous comment, my apologies! Just making sure this is clear.

**Comment by abred** [comment]
that's what I was saying, no worries. might help someone else having this problem.

**Comment by pemami4911** [comment]
Fixed in PR #8
